File: packages/agent/daemon/claudesidecar/assistantstream.go
package claudesidecar

import (
	"fmt"
)

type assistantSegmentState struct {
	messageID   string
	messageBase string
	kind        string // "assistant" or "thinking"
	source      string // "live" or "fallback"
	snapshot    string
	completed   bool
}

// AssistantStreamProjector mirrors assistantStream.ts: it turns streaming
// deltas and terminal content blocks into assistant/thinking events without
// double-emitting text that already streamed.
type AssistantStreamProjector struct {
	activeTurnID       func() string
	emit               Emitter
	currentMessageBase string
	sequence           int
	segmentsByKey      map[string]*assistantSegmentState
	segmentOrder       []string
	segmentKeyByIndex  map[string]string
}

func NewAssistantStreamProjector(activeTurnID func() string, emit Emitter) *AssistantStreamProjector {
	return &AssistantStreamProjector{
		activeTurnID:      activeTurnID,
		emit:              emit,
		segmentsByKey:     map[string]*assistantSegmentState{},
		segmentKeyByIndex: map[string]string{},
	}
}

func (p *AssistantStreamProjector) Reset() {
	p.currentMessageBase = ""
	p.sequence = 0
	p.segmentsByKey = map[string]*assistantSegmentState{}
	p.segmentOrder = nil
	p.segmentKeyByIndex = map[string]string{}
}

func (p *AssistantStreamProjector) SetMessageBase(messageID string) {
	p.currentMessageBase = messageID
}

func (p *AssistantStreamProjector) Start(index *int, kind string) {
	p.ensureLiveSegment(index, kind)
}

func (p *AssistantStreamProjector) AppendDelta(index *int, kind string, delta string) {
	p.emitDelta(p.ensureLiveSegment(index, kind), delta)
}

func (p *AssistantStreamProjector) CompleteIndex(index *int) bool {
	segment := p.segmentForIndex(index, "")
	if segment == nil {
		return false
	}
	p.completeSegment(index, segment, "")
	return true
}

func (p *AssistantStreamProjector) CompleteContent(kind string, messageBase string, content string, usedSegmentIDs map[string]struct{}) {
	if content == "" {
		return
	}
	if existing := p.segmentWithContent(kind, messageBase, content, usedSegmentIDs); existing != nil {
		p.completeSegment(nil, existing, content)
		usedSegmentIDs[existing.messageID] = struct{}{}
		return
	}
	base := messageBase
	if base == "" {
		base = p.activeTurnID()
	}
	segment := p.createSegment(kind, "fallback", base, nil)
	p.completeSegment(nil, segment, content)
	usedSegmentIDs[segment.messageID] = struct{}{}
}

func (p *AssistantStreamProjector) emitDelta(segment *assistantSegmentState, delta string) {
	if delta == "" {
		return
	}
	segment.snapshot += delta
	eventType := "assistant_delta"
	if segment.kind != "assistant" {
		eventType = "thinking_delta"
	}
	p.emit(eventType, "", map[string]any{
		"turnId":    p.activeTurnID(),
		"messageId": segment.messageID,
		"content":   delta,
		"snapshot":  segment.snapshot,
	})
}

func (p *AssistantStreamProjector) ensureLiveSegment(index *int, kind string) *assistantSegmentState {
	if existing := p.segmentForIndex(index, kind); existing != nil && !existing.completed {
		return existing
	}
	return p.createSegment(kind, "live", "", index)
}

func (p *AssistantStreamProjector) createSegment(kind string, source string, messageBase string, liveIndex *int) *assistantSegmentState {
	base := messageBase
	if base == "" {
		base = p.currentMessageBase
	}
	if base == "" {
		base = p.activeTurnID()
	}
	messageID := fmt.Sprintf("claude-sdk:%s:%s:%s:%d", kind, base, source, p.sequence)
	p.sequence++
	key := kind + ":" + messageID
	segment := &assistantSegmentState{
		messageID:   messageID,
		messageBase: base,
		kind:        kind,
		source:      source,
	}
	p.segmentsByKey[key] = segment
	p.segmentOrder = append(p.segmentOrder, key)
	if liveIndex != nil {
		p.segmentKeyByIndex[indexKey(kind, *liveIndex)] = key
	}
	return segment
}

func (p *AssistantStreamProjector) segmentForIndex(index *int, kind string) *assistantSegmentState {
	if index == nil {
		return nil
	}
	if kind != "" {
		key, ok := p.segmentKeyByIndex[indexKey(kind, *index)]
		if !ok {
			return nil
		}
		return p.segmentsByKey[key]
	}
	if key, ok := p.segmentKeyByIndex[indexKey("assistant", *index)]; ok {
		return p.segmentsByKey[key]
	}
	if key, ok := p.segmentKeyByIndex[indexKey("thinking", *index)]; ok {
		return p.segmentsByKey[key]
	}
	return nil
}

func (p *AssistantStreamProjector) completeSegment(index *int, segment *assistantSegmentState, fallbackText string) {
	if index != nil {
		delete(p.segmentKeyByIndex, indexKey(segment.kind, *index))
	}
	if segment.completed {
		return
	}
	tail := ""
	if fallbackText != "" && len(fallbackText) >= len(segment.snapshot) && fallbackText[:len(segment.snapshot)] == segment.snapshot {
		tail = fallbackText[len(segment.snapshot):]
	}
	if tail != "" && (segment.source != "fallback" || segment.snapshot != "") {
		p.emitDelta(segment, tail)
	} else if segment.snapshot == "" && fallbackText != "" {
		segment.snapshot = fallbackText
	}
	segment.completed = true
	if segment.snapshot == "" {
		return
	}
	eventType := "assistant_completed"
	if segment.kind != "assistant" {
		eventType = "thinking_completed"
	}
	p.emit(eventType, "", map[string]any{
		"turnId":    p.activeTurnID(),
		"messageId": segment.messageID,
		"content":   segment.snapshot,
	})
}

func (p *AssistantStreamProjector) segmentWithContent(kind string, messageBase string, content string, usedSegmentIDs map[string]struct{}) *assistantSegmentState {
	messageBases := map[string]struct{}{}
	for _, base := range []string{messageBase, p.currentMessageBase, p.activeTurnID()} {
		if base != "" {
			messageBases[base] = struct{}{}
		}
	}
	var tailCandidate *assistantSegmentState
	for _, key := range p.segmentOrder {
		segment := p.segmentsByKey[key]
		if segment == nil {
			continue
		}
		_, used := usedSegmentIDs[segment.messageID]
		_, baseMatches := messageBases[segment.messageBase]
		if segment.kind == kind && !used && baseMatches && segment.snapshot == content {
			return segment
		}
		if tailCandidate == nil && segment.kind == kind && !segment.completed && !used && baseMatches &&
			len(content) >= len(segment.snapshot) && content[:len(segment.snapshot)] == segment.snapshot {
			tailCandidate = segment
		}
	}
	return tailCandidate
}

func indexKey(kind string, index int) string {
	return fmt.Sprintf("%s:%d", kind, index)
}

File: packages/agent/claude-sdk-sidecar/src/assistantStream.ts
import type { ClaudeSDKSidecarEventEmitter } from "./protocol.ts";

export type AssistantSegmentKind = "assistant" | "thinking";

type AssistantSegmentState = {
  readonly messageId: string;
  readonly messageBase: string;
  readonly kind: AssistantSegmentKind;
  readonly source: "live" | "fallback";
  snapshot: string;
  completed: boolean;
  usageEmitted: boolean;
};

export class AssistantStreamProjector {
  private readonly activeTurnId: () => string;
  private readonly emit: ClaudeSDKSidecarEventEmitter;
  private currentMessageBase = "";
  private sequence = 0;
  private pendingUsage: Record<string, unknown> | undefined;
  private readonly segmentsByKey = new Map<string, AssistantSegmentState>();
  private readonly segmentKeyByIndex = new Map<string, string>();

  constructor(activeTurnId: () => string, emit: ClaudeSDKSidecarEventEmitter) {
    this.activeTurnId = activeTurnId;
    this.emit = emit;
  }

  reset(): void {
    this.currentMessageBase = "";
    this.sequence = 0;
    this.pendingUsage = undefined;
    this.segmentsByKey.clear();
    this.segmentKeyByIndex.clear();
  }

  setPendingUsage(usage: Record<string, unknown> | undefined): void {
    this.pendingUsage = usage;
  }

  // Re-emit already-completed assistant rows when transcript usage arrives
  // after content_block_stop, so payload.usage is not dropped.
  flushCompletedAssistantUsage(): void {
    if (!this.pendingUsage) {
      return;
    }
    for (const segment of this.segmentsByKey.values()) {
      if (
        segment.kind !== "assistant" ||
        !segment.completed ||
        !segment.snapshot ||
        segment.usageEmitted
      ) {
        continue;
      }
      this.emit({
        type: "assistant_completed",
        payload: {
          turnId: this.activeTurnId(),
          messageId: segment.messageId,
          content: segment.snapshot,
          usage: this.pendingUsage
        }
      });
      segment.usageEmitted = true;
    }
  }

  setMessageBase(messageId: string): void {
    this.currentMessageBase = messageId;
  }

  start(index: number | undefined, kind: AssistantSegmentKind): void {
    this.ensureLiveSegment(index, kind);
  }

  appendDelta(
    index: number | undefined,
    kind: AssistantSegmentKind,
    delta: string
  ): void {
    this.emitDelta(this.ensureLiveSegment(index, kind), delta);
  }

  completeIndex(index: number | undefined): boolean {
    const segment = this.segmentForIndex(index);
    if (!segment) {
      return false;
    }
    this.completeSegment(index, segment);
    return true;
  }

  completeContent(
    kind: AssistantSegmentKind,
    messageBase: string,
    content: string,
    usedSegmentIds: Set<string>
  ): void {
    if (!content) {
      return;
    }
    const existing = this.segmentWithContent(
      kind,
      messageBase,
      content,
      usedSegmentIds
    );
    if (existing) {
      this.completeSegment(undefined, existing, content);
      usedSegmentIds.add(existing.messageId);
      return;
    }
    const segment = this.createSegment(
      kind,
      "fallback",
      messageBase || this.activeTurnId()
    );
    this.completeSegment(undefined, segment, content);
    usedSegmentIds.add(segment.messageId);
  }

  failContent(
    kind: AssistantSegmentKind,
    messageBase: string,
    content: string,
    usedSegmentIds: Set<string>
  ): void {
    if (!content) {
      return;
    }
    const existing = this.segmentWithContent(
      kind,
      messageBase,
      content,
      usedSegmentIds
    );
    const segment =
      existing ??
      this.createSegment(kind, "fallback", messageBase || this.activeTurnId());
    this.completeSegment(undefined, segment, content, true);
    usedSegmentIds.add(segment.messageId);
  }

  private emitDelta(segment: AssistantSegmentState, delta: string): void {
    if (!delta) {
      return;
    }
    segment.snapshot += delta;
    this.emit({
      type: segment.kind === "assistant" ? "assistant_delta" : "thinking_delta",
      payload: {
        turnId: this.activeTurnId(),
        messageId: segment.messageId,
        content: delta,
        snapshot: segment.snapshot
      }
    });
  }

  private ensureLiveSegment(
    index: number | undefined,
    kind: AssistantSegmentKind
  ): AssistantSegmentState {
    const existing = this.segmentForIndex(index, kind);
    if (existing && !existing.completed) {
      return existing;
    }
    return this.createSegment(kind, "live", "", index);
  }

  private createSegment(
    kind: AssistantSegmentKind,
    source: "live" | "fallback",
    messageBase = "",
    liveIndex?: number
  ): AssistantSegmentState {
    const base = messageBase || this.currentMessageBase || this.activeTurnId();
    const messageId = `claude-sdk:${kind}:${base}:${source}:${this.sequence++}`;
    const key = `${kind}:${messageId}`;
    const segment: AssistantSegmentState = {
      messageId,
      messageBase: base,
      kind,
      source,
      snapshot: "",
      completed: false,
      usageEmitted: false
    };
    this.segmentsByKey.set(key, segment);
    if (typeof liveIndex === "number") {
      this.segmentKeyByIndex.set(indexKey(kind, liveIndex), key);
    }
    return segment;
  }

  private segmentForIndex(
    index: number | undefined,
    kind?: AssistantSegmentKind
  ): AssistantSegmentState | undefined {
    if (typeof index !== "number") {
      return undefined;
    }
    if (kind) {
      const key = this.segmentKeyByIndex.get(indexKey(kind, index));
      return key ? this.segmentsByKey.get(key) : undefined;
    }
    const key =
      this.segmentKeyByIndex.get(indexKey("assistant", index)) ??
      this.segmentKeyByIndex.get(indexKey("thinking", index));
    return key ? this.segmentsByKey.get(key) : undefined;
  }

  private completeSegment(
    index: number | undefined,
    segment: AssistantSegmentState,
    fallbackText = "",
    failed = false
  ): void {
    if (typeof index === "number") {
      this.segmentKeyByIndex.delete(indexKey(segment.kind, index));
    }
    if (segment.completed && !failed) {
      return;
    }
    const tail =
      fallbackText && fallbackText.startsWith(segment.snapshot)
        ? fallbackText.slice(segment.snapshot.length)
        : "";
    if (tail && (segment.source !== "fallback" || segment.snapshot)) {
      this.emitDelta(segment, tail);
    } else if (!segment.snapshot && fallbackText) {
      segment.snapshot = fallbackText;
    }
    segment.completed = true;
    if (!segment.snapshot) {
      return;
    }
    const payload: Record<string, unknown> = {
      turnId: this.activeTurnId(),
      messageId: segment.messageId,
      content: segment.snapshot
    };
    if (segment.kind === "assistant" && this.pendingUsage) {
      payload.usage = this.pendingUsage;
      segment.usageEmitted = true;
    }
    this.emit({
      type:
        segment.kind === "assistant"
          ? failed
            ? "assistant_failed"
            : "assistant_completed"
          : "thinking_completed",
      payload
    });
  }

  private segmentWithContent(
    kind: AssistantSegmentKind,
    messageBase: string,
    content: string,
    usedSegmentIds: Set<string>
  ): AssistantSegmentState | undefined {
    const messageBases = new Set(
      [messageBase, this.currentMessageBase, this.activeTurnId()].filter(
        Boolean
      )
    );
    let tailCandidate: AssistantSegmentState | undefined;
    for (const segment of this.segmentsByKey.values()) {
      if (
        segment.kind === kind &&
        !usedSegmentIds.has(segment.messageId) &&
        messageBases.has(segment.messageBase) &&
        segment.snapshot === content
      ) {
        return segment;
      }
      if (
        !tailCandidate &&
        segment.kind === kind &&
        !segment.completed &&
        !usedSegmentIds.has(segment.messageId) &&
        messageBases.has(segment.messageBase) &&
        content.startsWith(segment.snapshot)
      ) {
        tailCandidate = segment;
      }
    }
    return tailCandidate;
  }
}

function indexKey(kind: AssistantSegmentKind, index: number): string {
  return `${kind}:${index}`;
}
